pkg/controller/snapshot_controller_base.go

```diff
@@ -0,0 +1,439 @@
+			return
+		}
+		glog.V(5).Infof("enqueued %q for sync", objName)
+		ctrl.contentQueue.Add(objName)
+	}
+}
+
+// snapshotWorker processes items from snapshotQueue. It must run only once,
+// syncSnapshot is not assured to be reentrant.
+func (ctrl *csiSnapshotController) snapshotWorker() {
+	workFunc := func() bool {
+		keyObj, quit := ctrl.snapshotQueue.Get()
+		if quit {
+			return true
+		}
+		defer ctrl.snapshotQueue.Done(keyObj)
+		key := keyObj.(string)
+		glog.V(5).Infof("snapshotWorker[%s]", key)
+
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		glog.V(5).Infof("snapshotWorker: snapshot namespace [%s] name [%s]", namespace, name)
+		if err != nil {
+			glog.V(4).Infof("error getting namespace & name of snapshot %q to get snapshot from informer: %v", key, err)
+			return false
+		}
+		snapshot, err := ctrl.snapshotLister.VolumeSnapshots(namespace).Get(name)
+		if err == nil {
+			if ctrl.shouldProcessSnapshot(snapshot) {
+				// The volume snapshot still exists in informer cache, the event must have
+				// been add/update/sync
+				glog.V(4).Infof("should process snapshot")
+				ctrl.updateSnapshot(snapshot)
+			}
+			return false
+		}
+		if err != nil && !errors.IsNotFound(err) {
+			glog.V(2).Infof("error getting snapshot %q from informer: %v", key, err)
+			return false
+		}
+		// The snapshot is not in informer cache, the event must have been "delete"
+		vsObj, found, err := ctrl.snapshotStore.GetByKey(key)
+		if err != nil {
+			glog.V(2).Infof("error getting snapshot %q from cache: %v", key, err)
+			return false
+		}
+		if !found {
+			// The controller has already processed the delete event and
+			// deleted the snapshot from its cache
+			glog.V(2).Infof("deletion of vs %q was already processed", key)
+			return false
+		}
+		snapshot, ok := vsObj.(*crdv1.VolumeSnapshot)
+		if !ok {
+			glog.Errorf("expected vs, got %+v", vsObj)
+			return false
+		}
+		if ctrl.shouldProcessSnapshot(snapshot) {
+			ctrl.deleteSnapshot(snapshot)
+		}
+		return false
+	}
+
+	for {
+		if quit := workFunc(); quit {
+			glog.Infof("snapshot worker queue shutting down")
+			return
+		}
+	}
+}
+
+// contentWorker processes items from contentQueue. It must run only once,
+// syncContent is not assured to be reentrant.
+func (ctrl *csiSnapshotController) contentWorker() {
+	workFunc := func() bool {
+		keyObj, quit := ctrl.contentQueue.Get()
+		if quit {
+			return true
+		}
+		defer ctrl.contentQueue.Done(keyObj)
+		key := keyObj.(string)
+		glog.V(5).Infof("contentWorker[%s]", key)
+
+		_, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			glog.V(4).Infof("error getting name of snapshotData %q to get snapshotData from informer: %v", key, err)
+			return false
+		}
+		content, err := ctrl.contentLister.Get(name)
+		if err == nil {
+			// The volume still exists in informer cache, the event must have
+			// been add/update/sync
+			ctrl.updateContent(content)
```

Comment:
Should you skip Content that is for another CSI driver to save some memory and CPU?

Reply:
yes, we should add that check.

Reply:
Done